lib/derivation.ml

```diff
@@ -96,5 +96,36 @@ let rec derive context = function
         Subtype (tyvar_mfa, TyAny)] (* cannot be TyFun (.., ..) because the arity is unknown. give up *)
      in
      Ok (tyvar_mfa, Conj cs)
-  | other ->
-     Error (Printf.sprintf "unsupported type: %s" (show_expr other))
+  | Case (e, clauses) ->
+    (* translate pattern to expression *)
+    let rec pattern_to_expr = function
+      | PatVar v -> Var v
+      | PatTuple es -> Tuple (es |> List.map ~f:(fun e -> pattern_to_expr e)) in
+    (* Var(p) *)
+    let rec variables = function
+      | PatVar v -> [(v, new_tyvar ())]
+      | PatTuple es -> es |> List.map ~f:(fun e -> variables e)
+                          |> List.fold_left ~init:[] ~f:(fun a b -> List.append a b) in
+    derive context e >>= fun (ty_e_t, c_e) ->
+    let beta = new_tyvar () in
+    let results = clauses |> result_map_m ~f:(fun ((p_n, g_n), b_n) ->
+      let p_n_tyvars = variables p_n in
+      let p_n_expr = pattern_to_expr p_n in
+      (* A ∪ { ... } *)
+      let added_context = 
+        List.fold_left
+        ~f:(fun context (v, tyvar) -> Context.add (Context.Key.Var v) tyvar context)
+        ~init:context
+        p_n_tyvars in
+      (* |- p_n : alpha_n, c_p *)
+      derive added_context p_n_expr >>= fun(ty_alpha_n, c_p) ->
+      (* PAT *)
+      derive added_context g_n >>= fun(ty_g_n, c_g) ->
+      (* |- b_n : beta_n *)
+      derive added_context b_n >>= fun(ty_beta_n, c_b) ->
+        Ok (Conj [
+          Subtype (ty_g_n, TyUnion (TySingleton (Atom "true"), TySingleton (Atom "false")));
```

Comment:
Thank you for the fix!
Could you add an comment about the reason of difference from the original paper?
For example, "In the original paper, guards become `true` type. However, we consider guards to be `bool` type because `true` type constraint is too strong and not intuitive."

Reply:
@amutake I added the comment in the commit https://github.com/dwango/fialyzer/pull/59/commits/e6887202739fcfa4f4add4ab973b637320b87e32

Reply:
Thank you!